### ISSUE TITLE
Columns are tracked as toggling between null and their value

### ISSUE THREAD

**Issue by simonw** [simonw/git-history#33]
This is a really dumb bug:

<img width="1028" alt="511__item_version__10_rows_where_where__item___2" src="https://user-images.githubusercontent.com/9599/144757736-c027be56-b351-4d0e-b9fb-4cfdda271a47.png">

**Comment by simonw** [comment]
Here's why: https://github.com/simonw/git-history/blob/559f2de32343b48db9df4af590e32ce6f79fb641/git_history/cli.py#L358-L372

If I look at just the most previous record in the DB I'll see the `null` columns that I stored to represent that the item hadn't changed!

**Comment by simonw** [comment]
I introduced this bug when I was trying to get resume to work properly, for #25.

**Comment by simonw** [comment]
Prior to b9dc03e9b2d80164a1ddf48e2b23f1fc733c3bb0 I was keeping a full copy of the most recently seen version of each item in memory so that I could do these comparisons.

Two potential routes forward here:

1. Switch back to the in-memory `item_id_to_previous_version = {}` dictionary, and write code that attempts to re-create that dictionary from the data in the database if the database already exists.
2. Write a much smarter version of that SQL query that hydrates the most recent state by considering all previous versions in the database AND their `item_changed` tables so that it can tell the difference between literal `null` and nothing-changed `null`.

Both of these options have major drawbacks.

A very common planned usage pattern for this tool is for it to apply just the most recently tracked commits on an otherwise giant database. If it takes an hour to generate a DB file against 20,000 commits I want going from 20,000 to 20,001 to be a quick operation.

For that, the second option is better: no need to assemble a huge in-memory data structure of every item in the history if that single applied commit only references a few items.

The other case is when you terminate the script early and restart it at commit 10 out of 20,000 - for that, wouldn't it be better to hydrate everything into memory first?

Actually I don't think it would be better in that case (and it's a far less important case). So I should go with the second option.

I would only do the expensive hydrate-back-from-database process once per item, then I'd keep those items resident in RAM and update that version of them (maybe as part of a LRU cache).

**Comment by simonw** [comment]
Gonna need pretty robust tests for the code that I write that rehydrates an item based on its version history.

**Comment by simonw** [comment]
Found this bug while building a demo (#30) against https://github.com/dbreunig/511-events-history using this:

```
git-history file 511.db events.json --id id --convert '
data = json.loads(content)
for event in data["Events"]:
    # Remove noisy updated timestamps
    del event["extension"]["message-header"]
    del event["extension"]["details"]["eventElementDetails"]["event-element-times"]["update-time"]
    del event["extension"]["ivr-lastUpdateDttm"]
    event["event_subtypes"] = event["event_subtypes"]["event_subtype"]
    event["id"] = event["extension"]["event-reference"]["event-identifier"]
    yield event
' --wal
```

**Comment by simonw** [comment]
I'm going to write a utility function, `rehydrate(db, item_id, namespace="item")`.

**Comment by simonw** [comment]
Or... much, much simpler: use the `item` table and return the version from that! If I only ever need to start with the most recent version that should work just fine, and I can avoid all of this complexity.

**Comment by simonw** [comment]
I was still seeing this bug using the following code:
```python
updated_values = {
    key: value
    for key, value in item.items()
    if (key not in previous_item)
    or is_different(previous_item[key], value)
}
```
I used `dictdiffer` to try and figure out what was happening:
```
(Pdb) from dictdiffer import diff
(Pdb) pprint([l for l in diff(previous_item, item)])
[('change', 'event_subtypes', ('["Roadwork"]', ['Roadwork'])),
 ('change',
  'areas',
  ('{"area": {"name": "Santa Clara", "id": "5393021", "link": {"@rel": "self", '
   '"@href": "http://geonames.org/5393021/"}}}',
   {'area': {'id': '5393021',
             'link': {'@href': 'http://geonames.org/5393021/', '@rel': 'self'},
             'name': 'Santa Clara'}})),
 ('change',
  'geography',
  ('{"crs": {"type": "name", "properties": {"name": '
   '"urn:ogc:def:crs:EPSG::4326"}}, "type": "Point", "coordinates": '
   '[-122.096407, 37.393068]}',
   {'coordinates': [-122.096407, 37.393068],
    'crs': {'properties': {'name': 'urn:ogc:def:crs:EPSG::4326'},
            'type': 'name'},
    'type': 'Point'})),
 ('change',
  'source_type',
  ('{"@xmlns": "http://511.org/open511-extensions", "#text": "Caltrans"}',
   {'#text': 'Caltrans', '@xmlns': 'http://511.org/open511-extensions'})),
 ('change',
  'source_id',
  ('{"@xmlns": "http://511.org/open511-extensions", "#text": '
   '"P82HB-0010-2021-04-09-09:01:00"}',
   {'#text': 'P82HB-0010-2021-04-09-09:01:00',
    '@xmlns': 'http://511.org/open511-extensions'})),
 ('change',
  'roads',
  ('[{"name": "CA-82 N", "from": "El Monte Ave", "to": "N San Antonio Rd", '
   '"direction": "Northbound", "state": "SOME_LANES_CLOSED", "lane_type": '
   '{"@xmlns": "http://511.org/open511-extensions", "#text": "Lane"}, '
   '"lane_status": {"@xmlns": "http://511.org/open511-extensions", "#text": '
   '"closed"}, "road_advisory": {"@xmlns": '
   '"http://511.org/open511-extensions", "#text": "until 3 PM"}, "article": '
   '{"@xmlns": "http://511.org/open511-extensions", "#text": "between"}}]',
   [{'article': {'#text': 'between',
                 '@xmlns': 'http://511.org/open511-extensions'},
     'direction': 'Northbound',
     'from': 'El Monte Ave',
     'lane_status': {'#text': 'closed',
                     '@xmlns': 'http://511.org/open511-extensions'},
     'lane_type': {'#text': 'Lane',
                   '@xmlns': 'http://511.org/open511-extensions'},
     'name': 'CA-82 N',
     'road_advisory': {'#text': 'until 3 PM',
                       '@xmlns': 'http://511.org/open511-extensions'},
     'state': 'SOME_LANES_CLOSED',
     'to': 'N San Antonio Rd'}])),
 ('remove',
  '',
  [('_id', 9),
   ('_item_id', '499fc333b7dcaf715f3f07d939f9abb42ecf03f3'),
   ('_commit', 1),
   ('description', None)])]
(Pdb) 
```
The version I loaded from the database (the previous version) has had its lists/dicts turned into JSON strings, but the version I am comparing it against has not.

Confusing though, because I thought I had accounted for that in my `is_different()` function:

```python
def is_different(one, two):
    "Compares values, including lists and dictionaries"
    return jsonify_if_needed(one) != jsonify_if_needed(two)
```

**Comment by simonw** [comment]
Maybe it's a problem with key-ordering in the JSON versions?
```
(Pdb) jsonify_if_needed('{"area": {"name": "Santa Clara", "id": "5393021", "link": {"@rel": ' '"self", "@href": "http://geonames.org/5393021/"}}}')
'{"area": {"name": "Santa Clara", "id": "5393021", "link": {"@rel": "self", "@href": "http://geonames.org/5393021/"}}}'
(Pdb) jsonify_if_needed({'area': {'id': '5393021', 'link': {'@href': 'http://geonames.org/5393021/', '@rel': 'self'}, 'name': 'Santa Clara'}})
'{"area": {"id": "5393021", "link": {"@href": "http://geonames.org/5393021/", "@rel": "self"}, "name": "Santa Clara"}}'
```

**Comment by simonw** [comment]
I'm going to use my own version of `jsonify_if_needed` which orders the keys (original is at https://github.com/simonw/sqlite-utils/blob/213a0ff177f23a35f3b235386366ff132eb879f1/sqlite_utils/db.py#L3004-L3014)

**Comment by simonw** [comment]
Maybe the `item` table should also have `_version` and `_item_full_hash` columns, reflecting the correct values for the latest seen version?

At which point `item` really does become a materialized copy of the most recent record in `item_version` which makes conceptual sense to me.

**Comment by simonw** [comment]
It's still breaking on an example where this code:
```python
updated_values = {
    key: value
    for key, value in item_flattened.items()
    if (key not in previous_item)
    or previous_item[key] != value
}
```
Is failing to create a dictionary of changes for two items with these characteristics:
```
(Pdb) set(item_flattened.keys())
{'severity', 'source_type', 'created', 'id', 'event_subtypes', 'roads', 'source_id', 'headline', 'geography', 'areas', 'event_type', 'status'}
(Pdb) set(previous_item.keys())
{'severity', 'source_type', '_commit', 'created', 'id', 'event_subtypes', '_id', 'description', 'roads', 'source_id', 'headline', '_item_id', 'geography', 'areas', 'event_type', 'status'}
```
In this case it looks like the previous item had a `description` record but the most recent item no longer has that record. That's a change, so I would expect the code to produce a `"description": None` pair.

**Comment by simonw** [comment]
Got it! There's a `description` column in the database, but this record doesn't set it, so it ends up stored as `null` in the database.

But then when we inflate the previous item from the database we get a `"description": None` pair.

Then when we compare the previous item to the current item we don't spot any differences, because the current item doesn't have a value for `"description"` at all.

<img width="970" alt="Solving_git-history_item_changed_bug_-_Jupyter_Notebook" src="https://user-images.githubusercontent.com/9599/144790201-dc04ebcf-c7a9-4b94-920e-3e18aff25d90.png">

**Comment by simonw** [comment]
But then why did the hash change?

**Comment by simonw** [comment]
Relevant code from my laptop:
```python
                    if debug:
                        db["debug"].insert(
                            {
                                "hash": item_full_hash,
                                "content": json.dumps(item, default=repr, sort_keys=True),
                            },
                            pk="hash",
                            replace=True,
                        )

                    item_is_new = item_id not in item_id_to_last_full_hash
                    item_full_hash_has_changed = (
                        item_id_to_last_full_hash.get(item_id) != item_full_hash
                    )

                    updated_values = {}

                    if item_is_new or item_full_hash_has_changed:
                        # TODO: delete-me
                        previous_item_hash = item_id_to_last_full_hash.get(item_id)

                        # It's either new or the content has changed - so update item and insert an item_version
                        item_id_to_last_full_hash[item_id] = item_full_hash
                        version = item_id_to_version.get(item_id, 0) + 1
                        item_id_to_version[item_id] = version

                        previous_item = None
                        if not item_is_new:
                            previous_item = get_item(db, item_table, item_id)

                        # Add or update item
                        item_pk = db[item_table].lookup(
                            {"_item_id": item_id},
                            dict(item_flattened, _commit=commit_id),
                            column_order=("_id", "_item_id"),
                            foreign_keys=(("_commit", "commits", "id"),),
                            pk="_id",
                        )

                        if full_versions:
                            # Record full copies in item_version
                            item_version = dict(
                                item_flattened, _item=item_pk, _version=version, _commit=commit_id
                            )
                        else:
                            # Only record the columns that have changed
                            if previous_item is not None:
                                print("\n\nFiguring out updated values:\n", pformat(
                                    list(dictdiffer.diff(item_flattened, previous_item))))
                                updated_values = {
                                    key: value
                                    for key, value in item_flattened.items()
                                    if (key not in previous_item)
                                    or previous_item[key] != value
                                }
                            else:
                                updated_values = item_flattened

                            item_version = dict(
                                updated_values,
                                _item=item_pk,
                                _version=version,
                                _commit=commit_id,
                                _item_full_hash=item_full_hash,
                            )
```
The problem appears to be that the object corresponding to `previous_item_hash` - and hence the one we are comparing against to figure out `updated_values` - is incorrect. There HAS been a change between `previous_item` and `item` but it's not picked up by the `updated_values =` line.

Really confused here, not even able to write notes that make sense.

**Comment by simonw** [comment]
Next step: reduce it to pseudo-code, possibly break it into smaller functions, try again.

**Comment by simonw** [comment]
Just noticed this code here won't work with namespaces other than `item` - need to fix that: https://github.com/simonw/git-history/blob/a5cd18cf20c0204be637197815cc8fa286dab953/git_history/cli.py#L221-L235

**Comment by simonw** [comment]
I think I finally spotted the bug!

https://github.com/simonw/git-history/blob/ca9c8ebf703462352b1e4f36940f0cc4184a610b/git_history/cli.py#L334-L341

The comment says "add or update" but that's not what `.lookup()` does - if the first dictionary finds a match the second dictionary is ignored, rather than being used to update the row.

**Comment by simonw** [comment]
Yes, that was it! Replaced it with this instead (since `sqlite-utils` doesn't yet have an `upsert()` equivalent that works against a non-primary-key):
```python
# Add or update item
item_pk = db[item_table].lookup(
    {"_item_id": item_id},
    column_order=("_id", "_item_id"),
    foreign_keys=(("_commit", "commits", "id"),),
    pk="_id",
)
db[item_table].update(
    item_pk,
    dict(item_flattened, _item_id=item_id, _commit=commit_pk),
    alter=True,
)
```

**Comment by simonw** [comment]
Ran into a new problem with this commit: https://github.com/dbreunig/511-events-history/commit/a1cf2f7863a184884294578b4c944985d67c6320#diff-2d25bf4e2a1f8a3ddfc438667b0af9a206b2c660e452d9a5e4dbca0a004f8f4dL1703

<img width="1356" alt="Banners_and_Alerts_and_Latest_data__Mon_Apr_12_21_51_42_UTC_2021_·_dbreunig_511-events-history_a1cf2f7" src="https://user-images.githubusercontent.com/9599/144910023-fab8c366-9a91-4b3a-9eac-be09556bfad1.png">

The `closure_geometry` column was removed entirely. How should this be modeled in the database?

I think removed columns should be modeled the same as columns that have been set to `null` - so there should be a `null` in that column in the `item_version` record and a many-to-many `item_changed` record should show that the column was modified.

**Comment by simonw** [comment]
Merged PR #36, which fixes this issue.
